File: sdks/python/src/bigrag/_realtime.py
from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncGenerator
from typing import Any
from uuid import uuid4

from websockets.asyncio.client import connect

from bigrag.types.realtime import RealtimeMessage


class RealtimeConnection:
    def __init__(self, client) -> None:
        self._client = client
        self._socket = None
        self._reader: asyncio.Task | None = None
        self._queues: dict[str, asyncio.Queue[RealtimeMessage | None]] = {}

    async def connect(self) -> None:
        if self._socket is not None:
            return
        self._socket = await connect(
            _realtime_url(self._client.base_url),
            additional_headers=self._client._headers(),
            open_timeout=self._client.timeout,
        )
        self._reader = asyncio.create_task(self._read_loop())

    async def close(self) -> None:
        for queue in self._queues.values():
            queue.put_nowait(None)
        self._queues.clear()
        if self._reader is not None:
            self._reader.cancel()
            await asyncio.gather(self._reader, return_exceptions=True)
            self._reader = None
        if self._socket is not None:
            await self._socket.close()
            self._socket = None

    async def subscribe(
        self, topic: str, params: dict[str, Any] | None = None
    ) -> AsyncGenerator[RealtimeMessage, None]:
        await self.connect()
        subscription_id = uuid4().hex
        queue: asyncio.Queue[RealtimeMessage | None] = asyncio.Queue()
        self._queues[subscription_id] = queue
        await self._send(
            {
                "type": "subscribe",
                "id": subscription_id,
                "topic": topic,
                "params": _compact_params(params or {}),
            }
        )
        try:
            while True:
                message = await queue.get()
                if message is None:
                    return
                yield message
                if message.get("type") == "complete":
                    return
        finally:
            await self._send({"type": "unsubscribe", "id": subscription_id})
            self._queues.pop(subscription_id, None)
            queue.put_nowait(None)

    async def _send(self, message: dict[str, Any]) -> None:
        if self._socket is None:
            return
        await self._socket.send(json.dumps(message))

    async def _read_loop(self) -> None:
        try:
            while True:
                raw = await self._socket.recv()
                if isinstance(raw, bytes):
                    raw = raw.decode()
                try:
                    message = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                if not isinstance(message, dict):
                    continue
                if message.get("type") in {"heartbeat", "pong", "subscribed"}:
                    continue
                subscription_id = message.get("id")
                if not isinstance(subscription_id, str):
                    continue
                queue = self._queues.get(subscription_id)
                if queue is not None:
                    queue.put_nowait(message)
        finally:
            for queue in self._queues.values():
                queue.put_nowait(None)


class RealtimeResource:
    def __init__(self, client) -> None:
        self._client = client

    def connect(self) -> RealtimeConnection:
        return RealtimeConnection(self._client)

    async def subscribe(
        self, topic: str, params: dict[str, Any] | None = None
    ) -> AsyncGenerator[RealtimeMessage, None]:
        connection = self.connect()
        try:
            async for message in connection.subscribe(topic, params):
                yield message
        finally:
            await connection.close()


def _realtime_url(base_url: str) -> str:
    if base_url.startswith("https://"):
        return f"wss://{base_url.removeprefix('https://').rstrip('/')}/v1/realtime"
    if base_url.startswith("http://"):
        return f"ws://{base_url.removeprefix('http://').rstrip('/')}/v1/realtime"
    return f"{base_url.rstrip('/')}/v1/realtime"


def _compact_params(params: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in params.items() if value is not None}

File: sdks/typescript/src/realtime.ts
import { type RequestClient, USER_AGENT } from "./core.js";
import { APIConnectionError } from "./errors.js";
import type { RealtimeMessage } from "./types/index.js";

type AnySocket = {
  addEventListener?: (event: string, handler: (event: unknown) => void, options?: unknown) => void;
  close: () => void;
  on?: (event: string, handler: (...args: unknown[]) => void) => void;
  readyState?: number;
  send: (data: string) => void;
};

const OPEN = 1;

class AsyncQueue<T> {
  private readonly items: T[] = [];
  private readonly waiters: ((value: T | undefined) => void)[] = [];
  private closed = false;

  push(item: T): void {
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter(item);
      return;
    }
    this.items.push(item);
  }

  close(): void {
    this.closed = true;
    for (const waiter of this.waiters.splice(0)) waiter(undefined);
  }

  next(): Promise<T | undefined> {
    const item = this.items.shift();
    if (item !== undefined) return Promise.resolve(item);
    if (this.closed) return Promise.resolve(undefined);
    return new Promise((resolve) => this.waiters.push(resolve));
  }
}

export class BigRAGRealtimeConnection {
  private socket: AnySocket | null = null;
  private connecting: Promise<void> | null = null;
  private closing = false;
  private closedError: Error | null = null;
  private readonly queues = new Map<string, AsyncQueue<RealtimeMessage<unknown>>>();

  constructor(private readonly client: RequestClient) {}

  async connect(): Promise<void> {
    if (this.socket?.readyState === OPEN) return;
    if (this.connecting) return this.connecting;
    this.connecting = this.open();
    try {
      await this.connecting;
    } finally {
      this.connecting = null;
    }
  }

  async close(): Promise<void> {
    this.closing = true;
    for (const queue of this.queues.values()) queue.close();
    this.queues.clear();
    this.socket?.close();
    this.socket = null;
  }

  async *subscribe<T = unknown>(
    topic: string,
    params: Record<string, unknown> = {},
  ): AsyncGenerator<RealtimeMessage<T>> {
    await this.connect();
    const id = randomId();
    const queue = new AsyncQueue<RealtimeMessage<unknown>>();
    this.queues.set(id, queue);
    this.send({ type: "subscribe", id, topic, params: compactParams(params) });
    try {
      while (true) {
        const message = await queue.next();
        if (!message) {
          if (this.closedError) throw this.closedError;
          return;
        }
        yield message as RealtimeMessage<T>;
        if (message.type === "complete" || message.type === "error") return;
      }
    } finally {
      this.send({ type: "unsubscribe", id });
      this.queues.delete(id);
      queue.close();
    }
  }

  private async open(): Promise<void> {
    const { Socket, supportsHeaders } = await websocketConstructor(this.client.apiKey !== "");
    if (this.client.apiKey && !supportsHeaders) {
      throw new APIConnectionError(
        "Browser WebSocket cannot send the Authorization header required for API-key auth; use session-cookie auth or a collection realtime token.",
      );
    }
    const url = realtimeUrl(this.client.baseUrl);
    const headers: Record<string, string> = { "User-Agent": USER_AGENT };
    if (this.client.apiKey) headers.Authorization = `Bearer ${this.client.apiKey}`;
    const socket = (
      supportsHeaders ? new Socket(url, undefined, { headers }) : new Socket(url)
    ) as AnySocket;
    this.socket = socket;
    await new Promise<void>((resolve, reject) => {
      let settled = false;
      const done = (fn: () => void) => {
        if (settled) return;
        settled = true;
        fn();
      };
      on(socket, "open", () => done(resolve));
      on(socket, "error", (event) =>
        done(() => reject(new APIConnectionError(websocketErrorMessage(event)))),
      );
      on(socket, "message", (event) => this.handleMessage(event));
      on(socket, "close", () => this.handleClose());
    });
  }

  private send(message: unknown): void {
    if (!this.socket || this.socket.readyState !== OPEN) return;
    this.socket.send(JSON.stringify(message));
  }

  private handleMessage(event: unknown): void {
    const raw = messageData(event);
    if (raw === undefined) return;
    let message: RealtimeMessage<unknown>;
    try {
      message = JSON.parse(raw) as RealtimeMessage<unknown>;
    } catch {
      return;
    }
    if (message.type === "heartbeat" || message.type === "pong" || message.type === "subscribed") {
      return;
    }
    if (!message.id) return;
    this.queues.get(message.id)?.push(message);
  }

  private handleClose(): void {
    this.socket = null;
    if (!this.closing) {
      this.closedError = new APIConnectionError("Realtime connection closed unexpectedly");
    }
    for (const queue of this.queues.values()) queue.close();
    this.queues.clear();
  }
}

export class RealtimeResource {
  constructor(private readonly client: RequestClient) {}

  connect(): BigRAGRealtimeConnection {
    return new BigRAGRealtimeConnection(this.client);
  }

  async *subscribe<T = unknown>(
    topic: string,
    params: Record<string, unknown> = {},
  ): AsyncGenerator<RealtimeMessage<T>> {
    const connection = this.connect();
    try {
      yield* connection.subscribe<T>(topic, params);
    } finally {
      await connection.close();
    }
  }
}

const randomId = () => {
  const crypto = (globalThis as { crypto?: { randomUUID?: () => string } }).crypto;
  if (crypto?.randomUUID) return crypto.randomUUID();
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}-${Math.random().toString(36).slice(2)}`;
};

const websocketErrorMessage = (event: unknown): string => {
  if (event && typeof event === "object") {
    const candidate = event as { message?: unknown; error?: { message?: unknown } };
    if (typeof candidate.message === "string") return candidate.message;
    if (candidate.error && typeof candidate.error.message === "string") {
      return candidate.error.message;
    }
  }
  return "WebSocket connection failed";
};

const compactParams = (params: Record<string, unknown>) =>
  Object.fromEntries(Object.entries(params).filter(([, value]) => value !== undefined));

const realtimeUrl = (baseUrl: string) => {
  const url = new URL("/v1/realtime", `${baseUrl}/`);
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  return url.toString();
};

const websocketConstructor = async (needsHeaders: boolean) => {
  if (needsHeaders && !isBrowserRuntime()) return websocketModuleConstructor();
  const globalSocket = (globalThis as { WebSocket?: unknown }).WebSocket;
  if (globalSocket) {
    return {
      Socket: globalSocket as new (...args: unknown[]) => AnySocket,
      supportsHeaders: false,
    };
  }
  return websocketModuleConstructor();
};

const websocketModuleConstructor = async () => {
  const dynamicImport = new Function("specifier", "return import(specifier)") as (
    specifier: string,
  ) => Promise<{ WebSocket?: new (...args: unknown[]) => AnySocket; default?: unknown }>;
  const module = await dynamicImport("ws");
  return {
    Socket: (module.WebSocket ?? module.default) as new (...args: unknown[]) => AnySocket,
    supportsHeaders: true,
  };
};

const isBrowserRuntime = () =>
  (globalThis as { window?: unknown }).window !== undefined &&
  (globalThis as { window?: unknown }).window === globalThis;

const on = (socket: AnySocket, event: string, handler: (...args: unknown[]) => void) => {
  if (socket.addEventListener) {
    socket.addEventListener(event, (payload) => handler(payload));
    return;
  }
  socket.on?.(event, handler);
};

const messageData = (event: unknown): string | undefined => {
  if (typeof event === "string") return event;
  if (event instanceof Uint8Array) return new TextDecoder().decode(event);
  if (event && typeof event === "object" && "data" in event) {
    const data = (event as { data?: unknown }).data;
    if (typeof data === "string") return data;
    if (data instanceof Uint8Array) return new TextDecoder().decode(data);
  }
  return undefined;
};
